### PR TITLE
Added solutions for 3 challenges in English / data-structures 

### PR DESCRIPTION
**Pull request by sil3nthill**
- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

Closes  #35824

### CODE REVIEW

**Comment by sil3nthill** [comment]
is this PR ready to go ? if its okey can i start working on the challenges solution under arabic repo ?

**Comment by RandellDawson** [comment]
It may be a while before we get to this PR.  We have several others ahead of it, plus this PR will need to be reviewed by several mods to validate the solutions are correct (and that the existing tests are enough to validate them).

**Review by moT01 — CHANGES_REQUESTED** [review]
Can you format these a little better @sil3nthill - there's a lot of extra spaces and lines - the indentation should be two spaces for every level deeper you go

**Comment by sil3nthill** [comment]
@moT01  no problem i will change it now

**Comment by moT01** [comment]
Thanks @sil3nthill - there's still a lot of unnecessary empty lines in there - maybe I'm being too picky because they don't actually effect anything, but it makes the file look better if you get rid of them - also, the first solution changes the seed code - I'm thinking it would be best to keep all the seed code in the solution.

**Comment by sil3nthill** [comment]
Hello @moT01  can you check it now ? 
you asked to keep the same seed for the first one ! i didn't change the seed code

**Comment by moT01** [comment]
sorry @sil3nthill - what I meant was that the solution should be the seed code with the solution added. But, if I recall, one of the solutions had something different from what was in the seed code. Hope that makes sense. I can take a closer look a little later.

**Comment by sil3nthill** [comment]
hello @moT01  i took a close look at it , and yes i didn't change the seed code used in the solution , let me know if you want to edit anything else

**Review by moT01 — APPROVED** [review]
My mistake @sil3nthill - not sure what I was looking at when I thought that. But thanks for adding these and fixing it up. I think it looks good enough now.

**Review by Manish-Giri — CHANGES_REQUESTED** [review]
Hi @sil3nthill 

For the Adding an element at a specific index in a linked list challenge, please see the suggestions for your proposed solution - 
```js
this.addAt = function (index, element) {
    if (index > length || index < 0) {
      return false;
    }
    if (!head) {
      head = new Node(data);
    } else {
      var parent = null;
      var current = head;
      var i = 0;
      while (current && i < index) {
        parent = current;
        current = current.next;
        i++;
      }
      if (current) {
        var child = new Node(current.element);
        child.next = current.next;
        current.element = element;
        current.next = child;
        length++;
      } else {
        parent.next = new Node(element);
        length++;
      }
    }
  }
```

1. In a Linked List, there is usually no concept of a "parent" and a "child". All elements in the list are generally linked by references, and are considered contiguous to each other. "Parent" and "Child"  concepts are normally used in a Tree, or other hierarchical DS. Please refactor your variables to use `previous` or `previousNode` and `currentNode` (or something similar) to imply parent and child.

2. Also, in your `if` case, where you're checking if there is a `current` node, you can simplify your code a bit -
```js
   if (current) {
        var child = new Node(current.element);
        child.next = current.next;
        current.element = element;
        current.next = child;
        length++;
      }
```
- Ideally, if you have a reference to the "previous" node, and one to the "current" node, you need only two lines to add a new element between the two by just re-assigning the `next` references, something like this -
```js
 previousNode.next = newNode; 
 newNode.next = currentNode;
``` 

- I also see you're creating a new `Node` within both the `if` and the `else`. Since `element` in the newly created node will be the same irrespective of where the node is added (middle of the list or end of the list), you can create the new `Node` object outside the if/else block, and only change the `next` reference to this new object, within the if/else, as the case maybe.

- Instead of incrementing `length` in both if/else, you can do it *after* the if/else, and avoid code duplication.

**Comment by RandellDawson** [comment]
While testing the solution for `Remove Elements from a Linked List by Index` I noticed the current tests are not adequate and allow this invalid solution to pass.  For example, using the user's solution, if I execute the following code, I would expect the "dog" element to no longer be present, because it is the second element.  However, instead, it is still present but the size is correctly changed and the "dog" value is returned.

```js
var test = new LinkedList();
test.add('cat');
test.add('dog');
test.add('kitten');
test.add('bird');
console.log(JSON.stringify(test.head()));
console.log(test.size());
console.log();
console.log(test.removeAt(1));
console.log(JSON.stringify(test.head()));
console.log(test.size());
```
The actual result is:
```bash
{
  "element": "cat",
  "next": {
    "element": "dog",
    "next": {
      "element": "kitten",
      "next": {
        "element": "bird",
        "next": null
      }
    }
  }
}
linked list size: 4

removed element: dog
{
  "element": "cat",
  "next": {
    "element": "dog",
    "next": {
      "element": "kitten",
      "next": {
        "element": "bird",
        "next": null
      }
    }
  }
}
linked list size: 3
```
@sil3nthill You need to correct your solution. 

@Manish-Giri  @moT01 Should I create a new PR for the new tests needed or just create the needed test(s) and push as a new commit to this PR?

File: curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md [hunk 1]
```diff
@@ -112,6 +112,84 @@ function LinkedList() {
...
+      previousNode.next = currentNode.next;
+    }
+
+    length--;
+  };
+  this.removeAt = function (index) {
+    var currentNode = head;
+    var previous = head;
+    var count = 0;
+    if (index >= length || index < 0 || length == 0) {
```
Comment (RandellDawson) [inline curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md]:
```suggestion
    if (index >= length || index < 0) {
```

File: curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md [hunk 2]
```diff
@@ -112,6 +112,84 @@ function LinkedList() {
...
+      return null;
+    }
+    if (index === 0) {
+      previous = head;
```
Comment (RandellDawson) [inline curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md]:
```suggestion
      var removed = head.element;
```

File: curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md [hunk 3]
```diff
@@ -112,6 +112,84 @@ function LinkedList() {
...
+      head = currentNode.next;
```
Comment (RandellDawson) [inline curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md]:
```suggestion
      head = currentNode.next;
```

File: curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md [hunk 4]
```diff
@@ -112,6 +112,84 @@ function LinkedList() {
...
+      length--;
+      return head.element;
```
Comment (RandellDawson) [inline curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md]:
```suggestion
      return removed;
```

File: curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md [hunk 5]
```diff
@@ -112,6 +112,84 @@ function LinkedList() {
...
+    } else {
+      while (count < index - 1) {
```
Comment (RandellDawson) [inline curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md]:
```suggestion
      while (count <= index) {
```

**Review by RandellDawson — CHANGES_REQUESTED** [review]
@sil3nthill See the following changes which need to be made to your `removeAt` method which should resolve the issues I pointed out in my other comment.

@moT01 and @Manish-Giri @thecodingaviator I am also recommend that we delete the entire `remove` method from this challenge's seed code and solution as it is also not correct. If an element does not exist it errors out.

**Comment by moT01** [comment]
I would make a new issue to bring up your concerns @RandellDawson so we can try and just get this one merged - I didn't study the code, but the `remove` is added two lessons before this. It seems like a flaw in the function - rather than having to remove the whole function, it seems like there should be a catch for when the title doesn't exists.

**Comment by RandellDawson** [comment]
@moT01 The user's solution still needs to work as if already added the extra tests though, so it makes sense to me to go ahead and fix the tests.  Otherwise, I will still have to replace an invalid solution that got merged for some reason?  All I am trying to do is get the user's code to create a valid solution to the problem regardless of which tests are right or wrong.  The point is the solution may pass the current tests but it does not meet the challenge instructions requirements.

Regarding the remove function, I see no reason to have it since it is in no way used by the tests or is relevant to creating the `removeAt` method.  It is just taking up addition space.  I looked a many of the Binary Tree challenges and almost all of the "extra" methods are added as prototypes behind the scenes just for so the tests will work.

Also, I have created a PR #36121 to deal with the fact that the existing tests and solution are invalid for the `Remove Elements from a Linked List` challenge where the `remove` method was introduced.

**Comment by Manish-Giri** [comment]
I think the `remove(element)` method may have been added to the seed code so that campers could refer it while reading the description about `remove(element)` in the challenge instructions - 

> Just like our remove(element) method, we need to be careful not to orphan the rest of our list when we remove the node in our removeAt(index) method. We keep our nodes contiguous by making sure that the node that has reference to the removed node has a reference to the next node. 

But if this method is in no way being used by the tests, it should be removed. And even if someone somewhere down the line requires `remove(element)`, they could just call `removeAt(index)`, and add a few lines of code to find `index` of the given `element` by traversing the list. This avoids creating two methods which essentially do the same thing.

And then we could edit the challenge description to something like - 

> Just like our remove(element) method **_which we covered in a previous lesson_**, we need to be careful not to orphan the rest of our list when we remove the node in our removeAt(index) method. We keep our nodes contiguous by making sure that the node that has reference to the removed node has a reference to the next node.

File: curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md [hunk 6]
```diff
@@ -112,6 +112,84 @@ function LinkedList() {
...
+        previous = currentNode;
+        currentNode = currentNode.next;
+        count++;
+      }
+      var removed = previous.next;
+      previous.next = currentNode.next;
+      length--;
+      return removed.element;
```
Comment (RandellDawson) [inline curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md]:
```suggestion
      return removed;
```

**Comment by ojeytonwilliams** [comment]
@RandellDawson @Manish-Giri I went ahead and implemented your suggestions, since @sil3nthill has gone a bit quiet. I cleaned up `addAt` as much as I could and used @RandellDawson's suggestions as written, since there are some more changes coming in #36168

File: curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md [hunk 7]
```diff
@@ -9,7 +9,7 @@ challengeType: 1
...
 A common technique used to iterate through the elements of a linked list involves a <dfn>'runner'</dfn>, or sentinel, that 'points' at the nodes that your code is comparing. In our case, starting at the <code>head</code> of our list, we start with a <code>currentIndex</code> variable that starts at <code>0</code>. The <code>currentIndex</code> should increment by one for each node we pass.
-Just like our <code>remove(element)</code> method, we need to be careful not to orphan the rest of our list when we remove the node in our removeAt(index) method. We keep our nodes contiguous by making sure that the node that has reference to the removed node has a reference to the next node.
+Just like our <code>remove(element)</code> method, which we covered in a previous lesson, we need to be careful not to orphan the rest of our list when we remove the node in our removeAt(index) method. We keep our nodes contiguous by making sure that the node that has reference to the removed node has a reference to the next node.
```
Comment (RandellDawson) [inline curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md]: Below if my final suggestion based on how we have added links to previous challenges in other challenges.

```suggestion
Just like our <code>remove(element)</code> method, which <a href="learn/coding-interview-prep/data-structures/remove-elements-from-a-linked-list" target="_blank">we covered in a previous lesson</a>, we need to be careful not to orphan the rest of our list when we remove the node in our <code>removeAt(index)</code> method. We keep our nodes contiguous by making sure that the node that has reference to the removed node has a reference to the next node.
```

**Review by RandellDawson — CHANGES_REQUESTED** [review]
@ojeytonwilliams @Manish-Giri I believe the changes @ojeytonwilliams made have encompassed everything needed.  I do have one suggestion based on what we have done in other challenges which mentioned previous challenges.  Once this is committed, I would be fine approving/merging this PR.

File: curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md (hunk 7 above)
Comment (ojeytonwilliams) [reply curriculum/challenges/english/08-coding-interview-prep/data-structures/remove-elements-from-a-linked-list-by-index.english.md]: Seems good, I've pushed that change.

**Review by RandellDawson — APPROVED** [review]
I am fine with all the changes now.  @Manish-Giri How about you?  If you are OK with everything, then I say merge it.

**Review by Manish-Giri — APPROVED** [review]
LGTM 💯 

@ojeytonwilliams Thanks for adding in the suggestions!

**Pull request merged**